desc/objectives/_coils.py

```diff
@@ -96,102 +96,74 @@ def build(self, use_jit=True, verbose=1):  # noqa:C901
 
         """
         # local import to avoid circular import
-        from desc.coils import CoilSet, MixedCoilSet
+        from desc.coils import CoilSet, MixedCoilSet, _Coil
 
-        self._dim_f = 0
-        self._quad_weights = jnp.array([])
+        def _is_single_coil(c):
+            return isinstance(c, _Coil) and not isinstance(c, CoilSet)
 
-        def to_list(coilset):
-            """Turn a MixedCoilSet container into a list of what it's containing."""
-            if isinstance(coilset, list):
-                return [to_list(x) for x in coilset]
-            elif isinstance(coilset, MixedCoilSet):
-                return [to_list(x) for x in coilset]
+        def _prune_coilset_tree(coilset):
+            """Remove extra members from CoilSets (but not MixedCoilSets)."""
+            if isinstance(coilset, list) or isinstance(coilset, MixedCoilSet):
+                return [_prune_coilset_tree(c) for c in coilset]
             elif isinstance(coilset, CoilSet):
-                # use the same grid/transform for CoilSet
-                return to_list(coilset.coils[0])
+                # CoilSet only uses a single grid/transform for all coils
+                return _prune_coilset_tree(coilset.coils[0])
             else:
-                return [coilset]
+                return coilset  # single coil
+
+        coil = self.things[0]
+        grid = self._grid
+
+        # get individual coils from coilset
+        coils, structure = tree_flatten(coil, is_leaf=_is_single_coil)
+        self._num_coils = len(coils)
+
+        # map grid to list of length coils
+        if grid is None:
+            grid = [LinearGrid(N=2 * c.N + 5, endpoint=False) for c in coils]
+        if isinstance(grid, numbers.Integral):
+            grid = LinearGrid(N=self._grid, endpoint=False)
+        if isinstance(grid, _Grid):
+            grid = [grid] * self._num_coils
+        if isinstance(grid, list):
+            grid = tree_leaves(grid, is_leaf=lambda g: isinstance(g, _Grid))
+        assert len(grid) == len(coils)
```

Comment:
might want to make this a `ValueError` with a more descriptive error message?

Or maybe put a separate check after `if isinstance(grid, list):` since that's the only case where it could cause problems

Reply:
also what if the user passes in different grids for each coil in a regular `CoilSet`? I think in that case the code as is now will only use the first grid?

Reply:
Re your second question: yes the code will only use the first grid. That is how the existing code also behaves, so this PR is not changing that. My changes earlier this week corrected this, but then I reverted them because my solution didn't work for nested coilsets. We could try to fix this in the future.